### ISSUE TITLE
Story names are ambiguous when there are multiple stories in a benchmark with the same name but different grouping keys

### ISSUE THREAD

**Issue by zeptonaut** [catapult-project/catapult#3090]
This happens in the [loading.mobile benchmark](https://cs.chromium.org/chromium/src/tools/perf/benchmarks/loading.py?q=loading.py&sq=package:chromium&dr&l=39), which runs the same story twice: once without traffic shaping and once with 3G traffic shaping. When I tried to [kick off a bisect](https://bugs.chromium.org/p/chromium/issues/detail?id=673461), it failed because it erroneously assumed that I was trying to bisect on the story without traffic shaping, which was passing at both the good and bad revisions.

I'm not really sure of the correct thing to do here, but a couple of questions seem like they need to be answered:

1. Should we allow stories with the same name within a benchmark?
2. If so, how do we allow people to specify which of those stories they want to use for a given bisect?

My straw-man is that we should always ensure that there are no stories with the same name within a benchmark before running any benchmarks. We can leave it up to page set authors about how to differentiate between two similar stories that differ in key respects, like with traffic shaping. For the above example, I could certainly imagine the [loading_mobile.py](https://cs.chromium.org/chromium/src/tools/perf/page_sets/loading_mobile.py?q=loading.mobile&sq=package:chromium&dr=C&l=21) page set constructing stories with names like:

```
warm_cache:3g_connection:cnn.com
any_cache:ideal_connection:cnn.com
```

In other words, my vote is "don't add any new primitives to the system because we can do this with our existing primitives, but add an assertion to make sure people are doing the right thing". What do you all think?

/cc @nedn @eakuefner

**Comment by eakuefner** [comment]
Charlie, just to answer your questions, we already enforce that all stories in a story set should be unique up to name and grouping keys. So I feel like probably the right thing to do here is to also pass grouping keys to bisect.

**Comment by zeptonaut** [comment]
Ah - then we've already made the decision that the `(name, grouping key)` tuple has to be unique rather than just `(name)`.

Were all the implications of this thought through? It seems like there are a *lot*.

Anywhere that we currently do something based on the story name, we now need to also include the grouping key. For example:

1. In the dropdown where we show the "story name" when selecting a metric to display on a graph for Telemetry
2. When a specific story fails for a given benchmark on the waterfall, we currently show the name of the story that failed under the benchmark name on the builder run output page. For clarity, we should really show the benchmark name along with any grouping keys.
3. We need to add a `--grouping-key-filter` flag to Telemetry in addition to the `--story-filter` flag in order to make sure that we can accurately filter tests.
4. Sheriff-o-matic will need to know to look at grouping keys in addition to story names when bundling outages.
5. etc.

Especially for systems like Telemetry where there are so many connected systems, adding new primitives to systems is *really* expensive. What's the rationale behind why we did this instead of just either:

1. Automatically appending/prefixing any grouping keys onto the story name
2. Just keeping in place our constraint that all story names be unique, and letting pageset authors that have multiple same-named stories within a page set that only differ based on grouping keys (e.g. 3g traffic shaping, 4g traffic shaping) figure things out for themselves. For example, [it wouldn't have been too difficult for the pageset author of `loading_mobile` to just append the story names themselves](https://cs.chromium.org/chromium/src/tools/perf/page_sets/loading_mobile.py?q=pagecyclerstory&sq=package:chromium&l=126&dr=C).

**Comment by eakuefner** [comment]
Talked to @zeptonaut offline about the rationale; I think we agree that there are two next steps:

1. We need to make it so that the GTest test name Telemetry prints reflects the story grouping keys, and
2. We need to figure out why the structure of loading.mobile's chart JSON is different from page_cycler_v2's; we don't see the tir_label field polluted with per-page entries on PCv2.

**Comment by zeptonaut** [comment]
Hi Ethan,

Any update on this?

**Comment by eakuefner** [comment]
Hi Charlie, just coming back to this. I think the first change I describe is probably ready to be made modulo an email to telemetry-announce; I think addressing the second point requires a doc.

In particular, it seems that a post-processing pass on `StorySet` is necessary once all stories have been added (something like a `Finalize` method that must be called at latest in `StoryRunner` once all stories have been added but before stories begin being run) to either:

1. Add some kind of placeholder grouping key to stories that don't have a given grouping key that other stories do have, or
2. Blow up if not all stories in the set have each story grouping key set.

I'm probably in favor of alternative number 2, because it's more explicit, but I'd like to shop that around with folks. I'll write up the proposal today.

**Comment by zeptonaut** [comment]
Thanks for the update Ethan!
